File: math/greatest_common_divisor.go
package main

import "fmt"

func min(a, b int) int {
	if a <= b {
		return a
	}
	return b
}

// Time complecity Θ(min(a,b))
// Auxiliar space complexity Θ(1)
func gcd_naive(a, b int) int {
	if a == b {
		return a
	}
	res := min(a, b)
	for res > 0 {
		if a%res == 0 && b%res == 0 {
			break
		}
		res--
	}
	return res
}

// Time complecity Θ(min(a,b))
// Auxiliar space complexity Θ(min(a,b))
func gcd(a, b int) int {
	if a == b {
		return a
	} else if a > b {
		return gcd(a-b, b)
	} else {
		return gcd(a, b-a)
	}
}

// Time complecity Θ(log(min(a,b)))
// Auxiliar space complexity Θ(log(min(a,b)))
func gcd_improved(a, b int) int {
	if b == 0 {
		return a
	}
	return gcd_improved(b, a%b)
}

// Time Complexity: O(log(min(a,b))
// Auxiliary Space: O(log(min(a,b))
func least_common_multiple(a, b int) int { return a * b / gcd_improved(a, b) }

func main() {
	a := 12
	b := 28

	fmt.Println("Common divisor beteeen", a, b, "is", gcd_naive(a, b))
	fmt.Println("Common divisor beteeen", a, b, "is", gcd(a, b))
	fmt.Println("Common divisor beteeen", a, b, "is", gcd_improved(a, b))

	fmt.Println("Lest Common Multiple beteeen", a, b, "is", least_common_multiple(a, b))
}

File: math/greatest_common_divisor.c
#include <stdio.h>

int min(int a, int b) {
  if (a <= b) {
    return a;
  }
  return b;
}

// Time complecity Θ(min(a,b))
// Auxiliar space complexity Θ(1)
int gcd_naive(int a, int b) {
  if (a == b) {
    return a;
  }
  int res = min(a, b);
  while (res > 0) {
    if (a % res == 0 && b % res == 0) {
      break;
    }
    res--;
  }
  return res;
}

// Time complecity Θ(min(a,b))
// Auxiliar space complexity Θ(min(a,b))
int gcd(int a, int b) {
  if (a == b) {
    return a;
  } else if (a > b) {
    return gcd(a - b, b);
  } else {
    return gcd(a, b - a);
  }
}

// Time complecity Θ(log(min(a,b)))
// Auxiliar space complexity Θ(log(min(a,b)))
int gcd_improved(int a, int b) {
  if (b == 0) {
    return a;
  }
  return gcd_improved(b, a % b);
}

// Time Complexity: O(log(min(a,b))
// Auxiliary Space: O(log(min(a,b))
int least_common_multiple(int a, int b) { return a * b / gcd_improved(a, b); }

int main() {
  int a = 15;
  int b = 20;

  printf("Greatest common divisor beteeen %d and %d is %d\n", a, b,
         gcd_naive(a, b));
  printf("Greatest common divisor beteeen %d and %d is %d\n", a, b, gcd(a, b));
  printf("Greatest common divisor beteeen %d and %d is %d\n", a, b,
         gcd_improved(a, b));

  printf("Least common multiple beteeen %d and %d is %d\n", a, b,
         least_common_multiple(a, b));
}
